### ISSUE TITLE
'NoneType' object has no attribute 'send'

### ISSUE THREAD

**Issue by uphillbattle** [echoromeo/hanobo#24]
I occassionally loose the ability to control the Nobø devices. Please see error log below. Restarting Home Assistant resolves the problem - for a while - until next time...

```
Logger: homeassistant.components.websocket_api.http.connection
Source: custom_components/nobo_hub/climate.py:261
Integration: Home Assistant WebSocket API (documentation, issues)
First occurred: 10:44:26 (1 occurrences)
Last logged: 10:44:26

[139942307076704] 'NoneType' object has no attribute 'send'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 136, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1455, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1490, in _execute_service
    await handler.job.target(service_call)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 204, in handle_service
    await self.hass.helpers.service.entity_service_call(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 595, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 664, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 632, in _handle_entity_call
    await result
  File "/usr/src/homeassistant/homeassistant/components/climate/__init__.py", line 430, in async_set_hvac_mode
    await self.hass.async_add_executor_job(self.set_hvac_mode, hvac_mode)
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/nobo_hub/climate.py", line 223, in set_hvac_mode
    self.set_preset_mode(PRESET_NONE)
  File "/config/custom_components/nobo_hub/climate.py", line 261, in set_preset_mode
    self._nobo.create_override(mode, self._nobo.API.OVERRIDE_TYPE_CONSTANT, self._nobo.API.OVERRIDE_TARGET_ZONE, self._id)
  File "/usr/local/lib/python3.8/site-packages/pynobo.py", line 521, in create_override
    self.send_command(command)
  File "/usr/local/lib/python3.8/site-packages/pynobo.py", line 355, in send_command
    self.client.send(message + b'\r')
AttributeError: 'NoneType' object has no attribute 'send'
```

**Comment by uphillbattle** [comment]
A bit more information:

- The Nobø hub is not on the same network as HA
- I've set up a VPN connection so that the HA host can access the LAN hosting the Nobø hub
- This works well as long as the VPN tunnel is up
- Speculation: I'm wondering if network glitches (if the VPN tunnel goes down and then comes back up) lead to the above problem
- As long as the VPN tunnel is up, I can restart HA and everything works fine until it doesn't - can take hours or days

EDIT: Typo

**Comment by echoromeo** [comment]
I think what happens is that hanobo is trying to send a command while pynobo is disconnected from the hub.
pynobo is trying to reconnect at that point, but until that happens the self.client object is empty and calling the self.client.send function fails.
We will have to look into if this will require a fix in both pynobo and hanobo. Not sure if hanobo knows at any time that pynobo is disconnected.

**Comment by exxamalte** [comment]
I have the same issue and let the integration run with debug logging until the issue appeared:

```
2021-06-12 22:04:11 DEBUG (Thread-4) [pynobo] sending: ['HANDSHAKE']
2021-06-12 22:04:11 DEBUG (Thread-4) [pynobo] received: ['HANDSHAKE']
2021-06-12 22:04:26 DEBUG (Thread-4) [pynobo] sending: ['HANDSHAKE']
2021-06-12 22:04:26 DEBUG (Thread-4) [pynobo] received: ['HANDSHAKE']
2021-06-12 22:04:41 DEBUG (Thread-4) [pynobo] sending: ['HANDSHAKE']
2021-06-12 22:04:41 DEBUG (Thread-4) [pynobo] received: ['HANDSHAKE']
2021-06-12 22:04:56 DEBUG (Thread-4) [pynobo] sending: ['HANDSHAKE']
2021-06-12 22:04:56 DEBUG (Thread-4) [pynobo] received: ['HANDSHAKE']
2021-06-12 22:04:59 DEBUG (SyncWorker_17) [pynobo] Status at 2204 on weekday 5 is off
2021-06-12 22:04:59 DEBUG (SyncWorker_17) [pynobo] Current mode for zone Dining Room at 2204 is off
2021-06-12 22:04:59 DEBUG (SyncWorker_4) [pynobo] Current mode for zone Living Room at 2204 is comfort
2021-06-12 22:05:11 DEBUG (Thread-4) [pynobo] sending: ['HANDSHAKE']
2021-06-12 22:05:11 DEBUG (Thread-4) [pynobo] received: ['HANDSHAKE']
2021-06-12 22:05:26 DEBUG (Thread-4) [pynobo] sending: ['HANDSHAKE']
2021-06-12 22:05:26 DEBUG (Thread-4) [pynobo] received: ['HANDSHAKE']
2021-06-12 22:05:41 DEBUG (Thread-4) [pynobo] sending: ['HANDSHAKE']
2021-06-12 22:05:41 INFO (Thread-4) [pynobo] lost connection to hub ([Errno 104] Connection reset by peer)
2021-06-12 22:05:41 DEBUG (Thread-4) [pynobo] hub not rediscovered
2021-06-12 22:05:54 DEBUG (Thread-4) [pynobo] hub not rediscovered
2021-06-12 22:05:59 DEBUG (SyncWorker_6) [pynobo] Status at 2205 on weekday 5 is off
2021-06-12 22:05:59 DEBUG (SyncWorker_6) [pynobo] Current mode for zone Dining Room at 2205 is off
2021-06-12 22:05:59 DEBUG (SyncWorker_10) [pynobo] Current mode for zone Living Room at 2205 is comfort
2021-06-12 22:06:07 DEBUG (Thread-4) [pynobo] hub not rediscovered
2021-06-12 22:06:20 DEBUG (Thread-4) [pynobo] hub not rediscovered
2021-06-12 22:06:33 DEBUG (Thread-4) [pynobo] hub not rediscovered
2021-06-12 22:06:46 DEBUG (Thread-4) [pynobo] hub not rediscovered
2021-06-12 22:06:59 DEBUG (SyncWorker_14) [pynobo] Status at 2206 on weekday 5 is off
2021-06-12 22:06:59 DEBUG (SyncWorker_14) [pynobo] Current mode for zone Dining Room at 2206 is off
2021-06-12 22:06:59 DEBUG (SyncWorker_10) [pynobo] Current mode for zone Living Room at 2206 is comfort
2021-06-12 22:06:59 DEBUG (Thread-4) [pynobo] hub not rediscovered
2021-06-12 22:07:12 DEBUG (Thread-4) [pynobo] hub not rediscovered
2021-06-12 22:07:25 DEBUG (Thread-4) [pynobo] hub not rediscovered
2021-06-12 22:07:38 DEBUG (Thread-4) [pynobo] hub not rediscovered
```

So, essentially I see those `HANDSHAKE` messages every ~15 seconds, until one such handshake fails and after that I only see `hub not rediscovered` messages and lost control over my heaters, until I restart Home Assistant.

**Comment by uphillbattle** [comment]
My workaround here was to set up a local HA instance at the remote location (my cabin) where the Nobø system is. pynobo/hanobo is running on that local HA instance and since network glitches on the local LAN do not seem to occur, the local HA instance retains control of the Nobø system consistently.

At the other end of the above-mentioned VPN tunnel (at home) my Home HA instance connects to the Cabin HA instance using [Remote Home-Assistant](https://github.com/custom-components/remote_homeassistant) over the VPN tunnel. Remote Home-Assistant has not once lost the connection to the Cabin HA instance even though there has surely been VPN tunnel glitches. This way I have (so far) a reliable way of controlling the Cabin Nobø system from my Home HA instance - at the cost, of course, of another Rasberry Pi located at the cabin.

**Comment by exxamalte** [comment]
Thanks. In my case the Nobo Hub is in the same location as my Home Assistant server. Device and server are on different VLANs. I lose connectivity frequently - in the above example within just about 3 hours.

**Comment by uphillbattle** [comment]
Ah, ok, different setup then, and my workaround won't work easily for you. 

Anyway, I guess the root cause is that hanobo is not yet reliable enough when it comes to reestablishing the connection after a glitch.

**Comment by oyvindwe** [comment]
@exxamalte I have improved the reconnection logic in pynobo 1.2.0, which hopefully should fix this issue. However, you need a newer version of hanobo, either the one in PR #30, or wait for the official  integration to be released, see https://github.com/home-assistant/core/pull/50913

Probably root cause is (from the Nobø API documentation, page 4):

> From time to time the Hub will reboot (automatically approximately every 18 hours) or re-run DHCP to acquire a new IP address from the DHCP server on the network

**Comment by echoromeo** [comment]
@exxamalte and @uphillbattle, did the fix work? can we close this one?

**Comment by uphillbattle** [comment]
@echoromeo I run HA locally (to the Nobø hub) now (as [reported on 14 Jun](https://github.com/echoromeo/hanobo/issues/24#issuecomment-860588826)) and have no connection/reconnection issues. I experienced and reported this issue when my HA instance and the Nobø hub were on two different networks and HA accessed the Nobø hub over a VPN connection. I do appreciate the effort put in by you and @oyvindwe to fix the issue, but my workaround (an additional HA instance at the location of the Nobø hub) gives me additional benefits, so I'm sticking with that.

I will, however, see if I can set up the Nobø integration on my home HA and connect it directly over the VPN connection to check if the connection is stable. Will report back in a few days.

**Comment by uphillbattle** [comment]
I set up the Nobø integration on my home HA as advertised and configured it to communicate with the Nobø hub at my cabin (I have a VPN connection between my home and my cabin). This was the setup that gave me headaches when I reported it on 19 Febuary.

Now, however, the integration/connection has been stable without issues for 48 hours (since I last restarted HA). It appears that the original issue has been fixed. 🎉 Great work! 👏 

I believe the issue can be closed, so I'm closing it.
